pipeline/metadata/hyperquack_outcome.py

```diff
@@ -0,0 +1,202 @@
+  # Hyperquack v2 multiple errors
+  if error == "Status lines do not match": return _status_mismatch(received_status) # HTTP/S
+  if error == "Bodies do not match": return "body_mismatch" # HTTP/S
+  if "header field missing" in error: return "header_mismatch" # HTTP/S
+  if "header field mismatch" in error: return "header_mismatch" # HTTP/S
+  return None
+
+
+def _classify_hyperquack_error(error: Optional[str], scan_type: str,
+                               received_status: Optional[str]) -> str:
+  """
+  Args:
+    error string like 'read tcp 141.212.123.235:11397->117.78.42.54:9: read: connection reset by peer'
+    scan type: like 'echo, 'discard', etc
+
+  Returns:
+    outcome: string like "<stage>/<result>"
+  """
+  # Success
+  if error is None or error == "": return "expected/match"
+
+  if _is_system_failure(error):
+    return "setup/system_failure"
+
+  dial_error = _get_dial_error(error)
+  if dial_error: return f"dial/{dial_error}"
+
+  tls_failure = _get_tls_failure(error)
+  if tls_failure: return f"tls/{tls_failure}"
+
+  write_failure = _get_write_failure(error)
+  if write_failure: return f"write/{write_failure}"
+
+  read_failure = _get_read_failure(error, scan_type)
+  if read_failure: return f"read/{read_failure}"
+
+  http_failure = _get_http_content_verification_failure(error)
+  if http_failure: return f"http/{http_failure}"
+
+  mismatch_error = _get_content_mismatch(error, scan_type, received_status)
+  if mismatch_error: return f"content/{mismatch_error}"
+
+  return "unknown/unknown"
+
+# yapf: enable
+
+
+def classify_hyperquack_outcome(error: Optional[str], scan_type: str,
+                                received_status: Optional[str],
+                                matches_template: Optional[bool],
+                                is_known_blockpage: Optional[bool],
+                                blockpage_fingerprint: Optional[str],
+                                received_headers: List[str]) -> str:
+  """
+  Args:
+    error: string like "Incorrect web response: status lines don't match"
+    received_status: HTTP status like "200 OK"
+    matches_template: did the response match the control template
+    is_known_blockpage: did the page content match a blockpage
+    blockpage_fingerprint: which known blockpage did we match
+    received_headers: HTTP headers
+
+
+  Returns:
+    outcome string like "read/tcp.reset"
+  """
+  # TODO reorder this to more closely match when we see each signal
+
+  if is_known_blockpage and blockpage_fingerprint:
+    return "content/blockpage:" + blockpage_fingerprint
+
+  if _has_akamai_server_header(received_headers):
+    return "expected/trusted_host:akamai"
+
+  # Content mismatch for hyperquack v2 which didn't write
+  # content verification failures in the error field from 2021-04-26 to 2021-07-21
+  if not matches_template and (error is None or error == ""):
+    return "content/mismatch"
+
+  return _classify_hyperquack_error(
```

Comment:
I'd inline `_classify_hyperquack_error` here, so the entire high level logic is read in one place.

Reply:
done